checker-rs/src/notebook_client.rs

```diff
@@ -0,0 +1,351 @@
+use crate::NotebookUser;
+
+use enochecker::{
+    tokio::{
+        io::{AsyncBufRead, AsyncBufReadExt, AsyncWriteExt, BufStream},
+        net::TcpStream,
+    },
+    CheckerError, CheckerRequest, CheckerResult,
+};
+use std::io;
+
+use tracing::{debug, error, info, instrument, warn};
+/// reads until the delimiter given in delim is found or EOF if found or another Error occurs
+///
+/// # Return
+/// Number of bytes read
+pub async fn read_until_slice<'a, T: AsyncBufRead + Unpin>(
+    stream: &'a mut T,
+    delim: &'a [u8],
+    buf: &'a mut Vec<u8>,
+) -> io::Result<usize> {
+    let mut bytes_read: usize = 0;
+    let mut inc_read;
+
+    let delim_end = *delim.last().expect("DELIMITER must not be empty");
+
+    loop {
+        inc_read = stream.read_until(delim_end, buf).await?;
+        if inc_read == 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "Delimiter was not found",
+            ));
+        }
+        bytes_read += inc_read;
+        if bytes_read >= delim.len() && &buf[(buf.len() - delim.len())..buf.len()] == delim {
```

Comment:
`slice.ends_with()`?

Reply:
didn't know that existed, but yeah that should be it